apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-plugin/src/test/java/co/elastic/apm/agent/rabbitmq/RabbitMQTest.java

```diff
@@ -0,0 +1,672 @@
+        assertThat(transaction.getType()).isEqualTo("messaging");
+        assertThat(transaction.getNameAsString())
+            .isEqualTo("RabbitMQ message RECEIVE from %s", exchange.isEmpty() ? "<default>" : exchange);
+        assertThat(transaction.getFrameworkName()).isEqualTo("RabbitMQ");
+
+        checkMessage(transaction.getContext().getMessage(), exchange);
+    }
+
+    private static void checkMessage(Message message, String queueName) {
+        assertThat(message.getQueueName()).isEqualTo(queueName);
+
+        // RabbitMQ does not provide timestamp by default
+        assertThat(message.getAge()).isLessThan(0);
+    }
+
+
+    private static void checkMessageBodyNotCaptured(Message message) {
+        assertThat(message.getBodyForRead()).describedAs("body capture isn't supported").isNull();
+    }
+
+    private static final String[] DISTRIBUTED_TRACING_HEADERS = {
+        "elastic-apm-traceparent",
+        "tracestate",
+        "traceparent"
+    };
+
+    private static void checkDistributedTracingHeaders(Message message, boolean expectTracingHeaders) {
+        HashMap<String, String> headersMap = getHeadersMap(message);
+        if (expectTracingHeaders) {
+            assertThat(headersMap)
+                .describedAs("distributed tracing headers should be captured")
+                .containsKeys(DISTRIBUTED_TRACING_HEADERS);
+        } else {
+            assertThat(headersMap)
+                .describedAs("distributed tracing headers aren't expected")
+                .doesNotContainKeys(DISTRIBUTED_TRACING_HEADERS);
+        }
+    }
+
+    private static void checkHeaders(Message message, Map<String, String> expectedHeaders) {
+        HashMap<String, String> headersMap = getHeadersMap(message);
+        for (String key : DISTRIBUTED_TRACING_HEADERS) {
+            headersMap.remove(key);
+        }
+        assertThat(headersMap)
+            .describedAs("should contain entries of %s", expectedHeaders)
+            .containsAllEntriesOf(expectedHeaders);
+    }
+
+    private static HashMap<String, String> getHeadersMap(Message message) {
+        Headers headers = message.getHeaders();
+        HashMap<String, String> headersMap = new HashMap<>();
+        headers.forEach(h -> headersMap.put(h.getKey(), h.getValue().toString()));
+        return headersMap;
+    }
+
+    private static void checkSendSpan(Span span, String exchange) {
+        String expectedResource = exchange.isEmpty() ? "rabbitmq" : String.format("rabbitmq/%s", exchange);
+        checkSpanCommon(span,
+            "send",
+            String.format("RabbitMQ SEND to %s", exchange.isEmpty() ? "<default>" : exchange),
+            exchange,
+            expectedResource);
+
+    }
+
+    private static void checkPollSpan(Span span, String queue){
+        checkSpanCommon(span,
+            "poll",
+            String.format("RabbitMQ POLL from %s", queue),
+            queue,
+            "rabbitmq");
```

Comment:
Should contain the exchange in the resource name

Reply:
For polling, the exchange name is only known when there is a message, thus that would use two different names for the cases with/without a response, whereas it's the same polling operation. Also, I found it was more consistent with the "use the queue name when available, exchange name otherwise".

Reply:
My bad, I read it too quickly, here it's about the *resource name*, not the *span name*. Thus we should have:
- `rabbitmq` for polling when there is no message
- `rabbitmq/${exchange}` for polling when there is a message

We do not include queue name in resource name due to potentially high cardinality.

Also, that means that a successful message is being captured with queue name, and exchange name is visible in resource name, which could help to configure filtering using either queue name and exchange name.